ProcessOptimizer/model_systems/color_pH.py

```diff
@@ -0,0 +1,339 @@
+        print(f'DE2000: {round(DE2000, 4)}')
+
+    if test == True:
+        a1Prime = round(a1Prime, 4)
+        a2Prime = round(a2Prime, 4)
+        c1Prime = round(c1Prime, 4)
+        c2Prime = round(c2Prime, 4)
+        h1Prime = round(h1Prime, 4)
+        h2Prime = round(h2Prime, 4)
+        hBarPrime = round(hBarPrime, 4)
+        g = round(g, 4)
+        t = round(t, 4)
+        sL = round(sL, 4)
+        sC = round(sC, 4)
+        sH = round(sH, 4)
+        rT = round(rT, 4)
+        DE2000 = round(DE2000, 4)
+
+        return a1Prime, a2Prime, c1Prime, c2Prime, h1Prime, h2Prime, hBarPrime, g, t, sL, sC, sH, rT, DE2000
+
+    else:
+        return DE2000
+
+
+def find_closest_match(file_name, target_vector, target_columns):
+    # make docstring for this function
+    '''
+    file_name: name of the csv file
+    target_vector: vector of target values
+    target_columns: list of column names for the target values
+    
+    returns: row with the closest match to the target vector
+    '''
+    data = []
+    with open(file_name, 'r') as f:
+        reader = csv.reader(f)
+        headers = next(reader)
+        for row in reader:
+            data.append(row)
+
+    # Convert data to numpy array for easier manipulation
+    data = np.array(data)
+
+    # Get the indices of the target columns
+    target_indices = [headers.index(col) for col in target_columns]
+
+    # Extract the data for the target columns
+    target_data = data[:, target_indices].astype(float)
+
+    # Calculate the absolute differences for column 'B'
+    differences_B = np.abs(target_data[:, 0] - target_vector[0])
+
+    # Get the minimum difference for column 'B'
+    min_difference_B = np.min(differences_B)
+
+    # Get all rows where the difference for column 'B' is equal to the minimum difference
+    rows_with_min_difference_B = data[differences_B == min_difference_B]
+
+    # Extract the data for column 'C' from these rows
+    data_C = rows_with_min_difference_B[:, target_indices[1]].astype(float)
+
+    # Calculate the absolute differences for column 'C'
+    differences_C = np.abs(data_C - target_vector[1])
+
+    # Get the index of the minimum difference for column 'C'
+    min_difference_C_index = np.argmin(differences_C)
+
+    # Return the row with the minimum difference for column 'C'
+    return rows_with_min_difference_B[min_difference_C_index]
+
+
+def color_finder_dict(file_name, target_well):
+    '''
+    file_name: name of the csv file
+    target_well: well number of the target well
+    
+    returns: dictionary with the L,a,b values for the target well
+    '''
+    data = []
+    with open(file_name, 'r') as f:
+        reader = csv.reader(f)
+        headers = next(reader)
+        for row in reader:
+            data.append(row)
+
+    # Convert data to numpy array for easier manipulation
+    data = np.array(data)
+
+    # Get the indices of the target columns
+    target_indices = [headers.index(col) for col in ['Well','L','A','B']]
```

Comment:
`['Well','L','A','B']` could be an input to the function to make it more flexible. In the future, we might want to extract absorbance, or something else.

Let's keep it as is for now. We could add a comment, but we could also just remember that it is an option.